### ISSUE TITLE
[Bug]: dlib cnn_face_detection_model_v1 seems to use huge amount of memory

### ISSUE THREAD

**Issue by pkuskopf** [davisking/dlib#2846]
### What Operating System(s) are you seeing this problem on?

Windows

### dlib version

19.24.2

### Python version

3.10

### Compiler

 MSVC 19.36.32537.0

### Expected Behavior

I have installed the CPU version of dlib, and am trying to run the simple sample program "cnn_face_detector.py" from dlib.net, on a single image file with a single person's face. I have also installed the GPU version of dlib from source and attempted the same program.

### Current Behavior

As soon as the code reaches the line:
     `dets = cnn_face_detector(img, 1)`
in the CPU version of dlib:
     the memory working set explodes by 40GB, and the process starts generating a huge number of memory faults, causing it to run extremely slowly, and may take many minutes to process one file, despite the 32GB of memory in my PC, and:
in the GPU version of dlib:
     the process fails "cnn_face_detector Error while calling "cudaStreamSynchronize" (stream) ...  GPU_DATA.CPP Line 104 ... code: 2, reason: "out of memory"".

If I change the source code to use 0 up-sampling, then the memory usage is a bit smaller, but still significant, and the time taken is still significant.  Should this CNN model be so memory hungry and slow? It seems unusable at the moment.

### Steps to Reproduce

Created a minimal conda environment with just python 3.10, numpy, dlib, then ran the dlib.net/cnn_face_detector.py

**Comment by arrufat** [comment]
How big is the image?

**Comment by pkuskopf** [comment]
It's a standard photo image - 4476 x 6612 pixels. Total of 1.89 Mb. It takes many minutes.
But I've tried it out on other photos also, of about the same size (1.9 Mb) but with a large number of small faces and the runtime is only a few seconds. I haven't run enough different photos to be able to be sure about the correlation between runtime & memory usage, and whether they correlate to the photo composition, number of faces, etc.

**Comment by arrufat** [comment]
That is a pretty huge image.
The network will use an image pyramid, making the image 3.7 times larger.
You can read more about how this detector works [here](https://blog.dlib.net/2017/08/vehicle-detection-with-dlib-195_27.html). The network will also be converted to 32-bit floating-point numbers. This means the image will need about 1.3 GiB to be stored in memory.

After that, the image will go through the network, which has a bunch of convolutional and normalization layers of varying sizes. Each of those layers stores the result of their computation, making the memory usage explode.

TL;DR: it's not a bug, use a smaller image.

**Comment by pkuskopf** [comment]
@arrufat, 

Many thanks. Understood. 

However, would it be useful to resize the images, prior to running the detector? Using dlib.resize_image?

**Comment by arrufat** [comment]
Yes, that's what I'd do.

**Comment by pkuskopf** [comment]
Thanks, I'll close this
